pcap-linux.c

```diff
@@ -2661,14 +2661,17 @@ static int
 setup_mmapped(pcap_t *handle, int *status)
 {
 	struct pcap_linux *handlep = handle->priv;
-	int ret;
+	int ret, flags = MAP_ANONYMOUS | MAP_PRIVATE;
 
 	/*
 	 * Attempt to allocate a buffer to hold the contents of one
 	 * packet, for use by the oneshot callback.
 	 */
-	handlep->oneshot_buffer = malloc(handle->snapshot);
-	if (handlep->oneshot_buffer == NULL) {
+#ifdef MAP_32BIT
+	if (pcap_mmap_32bit) flags |= MAP_32BIT;
+#endif
+	handlep->oneshot_buffer = mmap(0, handle->snapshot, PROT_READ | PROT_WRITE, flags, -1, 0);
```

Comment:
Presumably `mmap()`, at least on Linux, supports non-multiple-of-the-page-size ''length'' arguments, with the mapping size being rounded up to a multiple of the page size.

Reply:
POSIX notes that "Use of *mmap*() may reduce the amount of memory available to other memory allocation functions."  Should it use `malloc()` if `pcap_mmap_32bit` isn't set and `mmap()` if it is?

Reply:
> Presumably `mmap()`, at least on Linux, supports non-multiple-of-the-page-size _length_ arguments, with the mapping size being rounded up to a multiple of the page size.

That's correct.

Reply:
> POSIX notes that "Use of _mmap_() may reduce the amount of memory available to other memory allocation functions." Should it use `malloc()` if `pcap_mmap_32bit` isn't set and `mmap()` if it is?

Perhaps. Large allocations with malloc() (i.e. those exceeding the page size) are backed by the same anonymous mappings, but because of the rounding we'd waste half a page on average. I'm not sure if saving that memory is worth complicating the code.